jbmc/src/java_bytecode/remove_instanceof.cpp

```diff
@@ -151,13 +170,49 @@ bool remove_instanceoft::lower_instanceof(
   // and less pessimistic aliasing assumptions possibly causing goto-symex to
   // explore in-fact-unreachable paths.
 
-  exprt subtype_check =
-    subtype_expr(object_clsid, target_name, class_hierarchy);
+  // In all cases require the input pointer is not null for any cast to succeed:
+
+  std::vector<exprt> test_conjuncts;
+  test_conjuncts.push_back(notequal_exprt(
+    check_ptr, null_pointer_exprt(to_pointer_type(check_ptr.type()))));
+
+  exprt object_class_identifier_field =
+    get_class_identifier_field(check_ptr, jlo, ns);
 
-  notequal_exprt not_null(
-    check_ptr, null_pointer_exprt(to_pointer_type(check_ptr.type())));
+  if(target_type_is_array)
+  {
+    test_conjuncts.push_back(equal_exprt(
+      object_class_identifier_field,
+      constant_exprt(JAVA_REFERENCE_ARRAY_CLASSID, string_typet())));
+
+    exprt object_array_dimension = get_array_dimension_field(check_ptr);
+    constant_exprt target_array_dimension = from_integer(
+      underlying_type_and_dimension.second, object_array_dimension.type());
+
+    if(underlying_type_is_object)
+    {
+      test_conjuncts.push_back(binary_relation_exprt(
+        object_array_dimension, ID_ge, target_array_dimension));
+    }
+    else
+    {
+      test_conjuncts.push_back(
+        equal_exprt(object_array_dimension, target_array_dimension));
+      test_conjuncts.push_back(subtype_expr(
+        get_array_element_type_field(check_ptr),
+        underlying_name,
+        class_hierarchy));
+    }
+  }
+  else if(!underlying_type_is_object)
+  {
+    test_conjuncts.push_back(subtype_expr(
+      get_class_identifier_field(check_ptr, jlo, ns),
+      underlying_name,
+      class_hierarchy));
+  }
 
-  expr = and_exprt(not_null, subtype_check);
+  expr = conjunction(test_conjuncts);
```

Comment:
:bulb: This logic seems like a great candidate for a unit test (that hopefully `simplify_expr` with constant strings will allow you to actually test the outcome).